### PR TITLE
add rate limiting for channel and user fetch

### PR DESCRIPTION
**Pull request by sprutner**

### CODE REVIEW

File: machine/clients/slack.py [hunk 1]
```diff
@@ -93,6 +94,80 @@ async def _process_users_channels(self, _: AsyncBaseSocketModeClient, req: Socke
             elif event["type"] == "member_joined_channel":
                 await self._on_member_joined_channel(event)
 
+    async def fetch_all_users(client):
```
Comment (DonDebonair) [inline machine/clients/slack.py:97]: Can you add type annotations for the function argument and return?

File: machine/clients/slack.py [hunk 2]
```diff
@@ -93,6 +94,80 @@ async def _process_users_channels(self, _: AsyncBaseSocketModeClient, req: Socke
...
+                    break
+
+            except Exception as e:
+                # Handle any other exceptions
+                logger.error(f"Unexpected error: {e}")
+                break
+
+        return all_users
+
+    async def fetch_all_channels(client) -> List[Dict[str, Any]]:
```
Comment (DonDebonair) [inline machine/clients/slack.py:131]: You can use lowercase `list` and `dict` here. In more recent versions of Python this is the recommended way

File: machine/clients/slack.py [hunk 3]
```diff
@@ -93,6 +94,80 @@ async def _process_users_channels(self, _: AsyncBaseSocketModeClient, req: Socke
...
+
+            except SlackApiError as e:
+                # Handle rate limiting
+                if e.response is 429:
+                    retry_after = int(e.response.headers.get("Retry-After", 1))
+                    logger.warning(f"Rate limit hit. Retrying after {retry_after} seconds...")
+                    await asyncio.sleep(retry_after)
+                else:
+                    # Handle other Slack API errors
+                    logger.error(f"Error fetching channels: {e.response['error']}")
```
Comment (DonDebonair) [inline machine/clients/slack.py:121]:
```suggestion
                    logger.error(f"Error fetching users: {e.response['error']}")
```

I would actually rethrow the exception here. If you don't, slack-machine is left with an incomplete list of users which will likely cause all kinds of undesired behavior. In my opionion, slack-machine shouldn't even start when failing to fetch users.

File: machine/clients/slack.py [hunk 4]
```diff
@@ -93,6 +94,80 @@ async def _process_users_channels(self, _: AsyncBaseSocketModeClient, req: Socke
...
+                    break
+
+            except Exception as e:
```
Comment (DonDebonair) [inline machine/clients/slack.py:124]: Why would we catch unknown exceptions? Let them bubble up and crash slack-machine. Easier to debug and as mentioned above, prevents undesired behavior

File: machine/clients/slack.py [hunk 5]
```diff
@@ -93,6 +94,80 @@ async def _process_users_channels(self, _: AsyncBaseSocketModeClient, req: Socke
...
+
+            except SlackApiError as e:
+                # Handle rate limiting
+                if e.response is 429:
+                    retry_after = int(e.response.headers.get("Retry-After", 1))
+                    logger.warning(f"Rate limit hit. Retrying after {retry_after} seconds...")
+                    await asyncio.sleep(retry_after)
+                else:
+                    # Handle other Slack API errors
+                    logger.error(f"Error fetching channels: {e.response['error']}")
```
Comment (DonDebonair) [inline machine/clients/slack.py:159]: Same as above, rethrow exception

File: machine/clients/slack.py [hunk 6]
```diff
@@ -93,6 +94,80 @@ async def _process_users_channels(self, _: AsyncBaseSocketModeClient, req: Socke
...
+                    break
+
+            except Exception as e:
```
Comment (DonDebonair) [inline machine/clients/slack.py:162]: same as above, don't catch exception

**Review by DonDebonair — CHANGES_REQUESTED** [review]
Hey @sprutner I assume you tested this on your workspace with many users?

I left a few comments to improve the code a bit. Great addition though! Excited to merge this in!

**Comment by sprutner** [comment]
Yeah I tested it on my workspace with like 68,000 users. It boots but it isn't getting any input from me as a user...perhaps workspace permissions issue. I pushed a couple commits but haven't tested it. I never set the python env up it was just a late night hack. I can look more into cleaning this up over the weekend :)

**Comment by DonDebonair** [comment]
Yeah it would help if you can do a bit of serious testing :)

**Comment by sprutner** [comment]
@DonDebonair doesn't seem like your tests run on each commit. It is hard for me to see if it's passing your tests here but re-requesting review. Installed the pre-commit hooks and ran them and got them to pass.

**Comment by DonDebonair** [comment]
Can you allow me to push to this branch on your fork? It's probably easier for me to fix the MyPy issues and I want to rebase the branch on `main` as well

**Comment by sprutner** [comment]
@DonDebonair done

**Comment by jasonwbarnett** [comment]
@sprutner alright to close?

**Comment by sprutner** [comment]
I think so. I forked this. I have been busy with other projects but when I
come back to this I can make another PR

All Good Things,
Seth Rutner

**Comment by DonDebonair** [comment]
@sprutner no need to open a new PR. #1194 done by @jasonwbarnett includes your code.

**Pull request closed without merging**